django/contrib/admin/views/main.py

```diff
@@ -303,14 +303,65 @@ def get_ordering(self, request, queryset):
         # Add the given query's ordering fields, if any.
         ordering.extend(queryset.query.order_by)
 
-        # Ensure that the primary key is systematically present in the list of
-        # ordering fields so we can guarantee a deterministic order across all
-        # database backends.
-        pk_name = self.lookup_opts.pk.name
-        if {'pk', '-pk', pk_name, '-' + pk_name}.isdisjoint(ordering):
-            # The two sets do not intersect, meaning the pk isn't present. So
-            # we add it.
-            ordering.append('-pk')
+        return self._get_deterministic_ordering(ordering)
+
+    def _get_deterministic_ordering(self, ordering):
+        """
+        Ensure a deterministic order across all database backends. Search for a
+        single field or unique together set of fields providing a total
+        ordering. If these are missing, augment the ordering with a descendant
+        primary key.
+        """
+        ordering = list(ordering)
+        ordering_fields = set()
+        total_ordering_fields = {'pk'}
+        total_ordering_fields.update(
+            field.attname for field in self.lookup_opts.fields
+            if field.unique and not field.null
+        )
+        for part in ordering:
+            # Search for single field providing a total ordering.
+            field_name = None
+            if isinstance(part, str):
+                if part.startswith('-'):
+                    field_name = part[1:]
+                else:
+                    field_name = part
```

Comment:
May as well do the following as a field name can only legally have a single `-` at the start:
```python
field_name = part.lstrip('-')
```